### PR TITLE
[Load] Fix bug of wrong file group aggregation when handling broker load job

### PR DESCRIPTION
**Pull request by morningman**
**Describe the bug**

**First**, In the broker load, we allow users to add multiple data descriptions. Each data description
 represents a description of a file (or set of files). Including file path, delimiter, table and 
partitions to be loaded, and other information.

When the user specifies multiple data descriptions, Doris currently aggregates the data 
descriptions belonging to the same table and generates a unified load task.

The problem here is that although different data descriptions point to the same table, 
they may specify different partitions. Therefore, the aggregation of data description
 should not only consider the table level, but also the partition level.

Examples are as follows:

data description 1 is: 
```
DATA INFILE("hdfs://hdfs_host:hdfs_port/input/file1")
INTO TABLE `tbl1`
PARTITION (p1, p2)
```

data description 2 is:
```
DATA INFILE("hdfs://hdfs_host:hdfs_port/input/file2")
INTO TABLE `tbl1`
PARTITION (p3, p4)
```
What user expects is to load file1 into partition p1 and p2 of tbl1, and load file2 into paritition
p3 and p4 of same table. But currently, it will be aggregated together, which result in loading
file1 and file2 into all partitions p1, p2, p3 and p4.

**Second**, the following 2 data descriptions are not allowed:

```
DATA INFILE("hdfs://hdfs_host:hdfs_port/input/file1")
INTO TABLE `tbl1`
PARTITION (p1, p2)
DATA INFILE("hdfs://hdfs_host:hdfs_port/input/file2")
INTO TABLE `tbl1`
PARTITION (p2, p3)
```

They have overlapping partition(p2), which is not support yet. And we should throw an Exception
to cancel this load job.

**Third**, there is a problem with the code implementation. In the constructor of 
`OlapTableSink.java`, we pass in a string of partition names separated by commas. 
But at the `OlapTableSink` level, we should be able to pass in a list of partition ids directly,
 instead of names.

ISSUE: #2823

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/load/PullLoadSourceInfo.java [hunk 1]
```diff
@@ -1,119 +0,0 @@
-// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/PullLoadSourceInfo.java]: This class has been consistent in the old version. Maybe you should keep it .

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    @Override
+    public void write(DataOutput out) throws IOException {
+        // The pull load source info doesn't need to be persisted.
+        // It will be recreated by origin stmt in prepare of load job.
+        // write 0 just for compatibility
+        out.writeInt(0);
+    }
+
+    public void readFields(DataInput in) throws IOException {
+        in.readInt();
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: Maybe in is a PullLoadSourceInfo??

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+ *  FileGroupAggKey(tbl1, [p3]) => List(file2);
+ *  FileGroupAggKey(tbl1, [p2]) => List(file1, file2);
+ *  
+ *  Although this transformation can be done automatically by system, but it change the "max_filter_ratio".
+ *  So we have to let user decide what to do.
+ */
+public class BrokerFileGroupAggInfo implements Writable {
+    private static final Logger LOG = LogManager.getLogger(BrokerFileGroupAggInfo.class);
+
+    private Map<FileGroupAggKey, List<BrokerFileGroup>> aggKeyToFileGroups = Maps.newHashMap();
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: Maybe Map<tableId, List<BrokerFileGroup>> is useful.

**Comment by EmmyMiao87** [comment]
If the use has two different partition info with the same table such as  file1->p1,p3, file2->p1,p2 , user should submit the load job such as file1->p1 , file1->p3, file2->p1, file2->p2.
In this situation, there are four loading task instead of two of them. 
Why don't you use the old struct such as Map<tableId, List<LoadInfo>>?

**Comment by EmmyMiao87** [comment]
The `PullLoadSourceInfo` has been consistent in some old version. I am afraid of the compatibility problems. Same as 'PullLoadPendingTask'

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroup.java [hunk 4]
```diff
@@ -77,8 +77,8 @@
     // filter the data which has been conformed
     private Expr whereExpr;
 
-    // Used for recovery from edit log
-    private BrokerFileGroup() {
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/src/main/java/org/apache/doris/load/BrokerFileGroup.java]: OK

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+ *  DATA INFILE("hdfs://hdfs_host:hdfs_port/input/file2")
+ *  INTO TABLE `tbl1`
+ *  PARTITION (p2)
+ *  
+ *  they will be aggregate like:
+ *  FileGroupAggKey(tbl1, [p1]) => List(file1);
+ *  FileGroupAggKey(tbl1, [p3]) => List(file2);
+ *  FileGroupAggKey(tbl1, [p2]) => List(file1, file2);
+ *  
+ *  Although this transformation can be done automatically by system, but it change the "max_filter_ratio".
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: No, it's not. for example
Assume that file1 should be loaded to p1 and p2,
and file2 should be loaded to p2 and p3.

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+        @Override
+        public String toString() {
+            StringBuilder sb = new StringBuilder();
+            sb.append("[").append(tableId).append(": ").append(partitionIds).append("]");
+            return sb.toString();
+        }
+    }
+
+    public BrokerFileGroupAggInfo() {
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: OK

File: fe/src/main/java/org/apache/doris/load/PullLoadSourceInfo.java (hunk 1 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/load/PullLoadSourceInfo.java]: It ok, I check it, it does no use at all now. and the old load job are gone before Doris version 0.11.

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java (hunk 2 above)
Comment (morningman) [reply fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: It won't, we already solve this in version 0.11

File: fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java [hunk 7]
```diff
@@ -224,7 +224,15 @@ public void write(DataOutput out) throws IOException {
     }
 
     public void readFields(DataInput in) throws IOException {
-        in.readInt();
+        int mapSize = in.readInt();
+        // just for compatibility, the following read objects are useless
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/BrokerFileGroupAggInfo.java]: A warning log is better ~

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**